### DISCUSSION TITLE
Any thoughts about components?

### DISCUSSION THREAD

**Discussion by azazel75** [ogrodnek/pyview#80] [General]
What's your idea of reusability across views? any thoughts on something similar to https://hexdocs.pm/phoenix_live_view/Phoenix.Component.html?

**Comment by ogrodnek** [comment]
@azazel75 Yes, I'd definitely love to have a component and live component implementation, just haven't gotten to it yet...

Even shorter term, template includes (which don't currently work) are high on the list of things to implement.

**Comment by azazel75** [comment]
That's a bummer! The custom templates are needed because of the diffing protocol, like in the original LiveView?

**Comment by ogrodnek** [comment]
Yup, that's right, we need to generate a tree of statics and dynamics for the diffs and to integrate with the client-side JS (I'm using the same Phoenix Liveview client side JS).

The library I'm currently using as the basis (https://github.com/dmulholl/ibis), does support includes already (for generating strings), I just haven't yet hacked in the statics/dynamics support.

I'm hoping if https://peps.python.org/pep-0501/ get implemented, it would be easier to also generate the tree we needed from python code (in addition to templating).

Theoretically, we could support another templating library if the library had support for generating some kind of parse tree instead of just output text...

**Comment by ogrodnek** [comment]
Quick update on this, I did add some initial support for template includes (you can see an example here: https://github.com/ogrodnek/pyview/blob/main/examples/views/includes/includes.html#L1 ), and more details on template includes in the [ibis template include docs](http://www.dmulholl.com/docs/ibis/master/tags.html#include), so now there is a mechanism to reuse markup across views.

I do still plan to add support for "live components" (Ala phoenix LiveView https://hexdocs.pm/phoenix_live_view/Phoenix.LiveComponent.html ) which encapsulate state in addition to markup and can handle events at some point...

**Comment by ogrodnek** [comment]
It's been awhile! My current thoughts on this is python 3.14 tstring support is going to make liveview templating much easier!

I have a work in progress branch on adding tstring template support: https://github.com/ogrodnek/pyview/pull/77

maybe even eventually integrating w/ something like: https://github.com/t-strings/tdom for better HTML/file support.

I think this will really open the door in terms of re-usability in python code and make "live components" easier to implement...

**Comment by ogrodnek** [comment]
Ok, we finally have some component support! I'd consider it "alpha", so, lmk if any feedback/bugs/suggestions.

Note that components are only supported when using the new tstring templates (which do require python 3.14).

see here for the release notes: https://github.com/ogrodnek/pyview/discussions/108

and docs here: https://pyview.rocks/core-concepts/live-components/
